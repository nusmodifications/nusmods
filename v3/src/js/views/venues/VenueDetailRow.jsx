// @flow
import React, { PureComponent } from 'react';
import { flatMap, noop } from 'lodash';
import { arrangeLessonsForWeek } from 'utils/timetables';
import { colorLessonsByKey } from 'utils/colors';
import Timetable from 'views/timetable/Timetable';

import type { DayAvailability, VenueLesson } from 'types/venues';
import type { Venue } from 'types/modules';

import styles from './VenueDetailRow.scss';

type Props = {
  name: Venue,
  availability: DayAvailability[],
  expanded: boolean,
  onClick: (Venue, string) => void,
  rootElementRef?: (?HTMLElement) => void, // For parent components to obtain a ref to the root HTMLElement
}

export default class VenueDetailRow extends PureComponent<Props> {
  static defaultProps = {
    name: '',
    availability: [],
    expanded: false,
    onClick: noop,
  };

  arrangedLessons() {
    if (!this.props.expanded) {
      return null;
    }

    const availability: DayAvailability[] = this.props.availability;
    const lessons = flatMap(availability, (day): VenueLesson[] => day.Classes)
      .map(venueLesson => ({ ...venueLesson, ModuleTitle: '' }));
    const coloredLessons = colorLessonsByKey(lessons, 'ModuleCode');
    return arrangeLessonsForWeek(coloredLessons);
  }

  render() {
    const { name, onClick } = this.props;
    const lessons = this.arrangedLessons();
    const venueHref = `/venues/${encodeURIComponent(name)}`;
    const rootElementRef: Function = this.props.rootElementRef || (() => {});

    return (
      <li className={styles.venueDetailRow} ref={rootElementRef}>
        <a
          href={venueHref}
          onClick={(e) => {
            e.preventDefault();
            onClick(name, venueHref);
          }}
        ><h4>{name}</h4></a>
        {lessons && (
          <div className={styles.venueTimetable}>
            <Timetable lessons={lessons} />
          </div>
        )}
      </li>
    );
  }
}
