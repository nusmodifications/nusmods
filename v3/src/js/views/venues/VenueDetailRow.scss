@import "~styles/utils/modules-entry.scss";

.venueDetailRow {
  padding-top: 1rem;
  border-top: 1px $gray-lighter solid;

  &:first-child {
    padding-top: 0.25rem;
    border-top: 0;
  }

  a {
    width: 100%;
  }
}

.venueTimetable {
  margin-bottom: 1rem;
  overflow: auto;
}
