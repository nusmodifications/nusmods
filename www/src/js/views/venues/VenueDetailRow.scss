@import "~styles/utils/modules-entry.scss";

.venueDetailRow {
  padding-top: 1rem;
  border-top: 1px var(--gray-lighter) solid;

  &:first-child {
    padding-top: 0.25rem;
    border-top: 0;
  }
}

.venueTimetable {
  composes: scrollable from global;
  margin-bottom: 1rem;
}
