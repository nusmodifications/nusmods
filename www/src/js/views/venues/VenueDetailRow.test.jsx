// @flow
import React from 'react';
import { shallow } from 'enzyme';
import venueInfo from '__mocks__/venueInformation.json';
import VenueDetailRow from './VenueDetailRow';

const minProps = {
  name: 'CQT/SR0622',
  availability: venueInfo['CQT/SR0622'],
};

describe('VenueDetailRow', () => {
  test('it displays an anchor tag when not expanded', () => {
    const wrapper = shallow(<VenueDetailRow {...minProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('it displays an anchor tag and a timetable when expanded', () => {
    const wrapper = shallow(<VenueDetailRow {...minProps} expanded />);
    expect(wrapper).toMatchSnapshot();
  });
});
