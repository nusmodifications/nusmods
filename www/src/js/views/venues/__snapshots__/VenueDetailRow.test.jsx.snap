// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VenueDetailRow it displays an anchor tag and a timetable when expanded 1`] = `
<li
  className="venueDetailRow"
>
  <a
    href="/venues/CQT%2FSR0622"
    onClick={[Function]}
  >
    <h4>
      CQT/SR0622
    </h4>
  </a>
  <div
    className="venueTimetable"
  >
    <Timetable
      isScrolledHorizontally={false}
      isVerticalOrientation={false}
      lessons={
        Object {
          "Thursday": Array [
            Array [
              Object {
                "ClassNo": "ST4",
                "DayText": "Thursday",
                "EndTime": "1500",
                "LessonType": "Tutorial",
                "ModuleCode": "PC2132",
                "ModuleTitle": "",
                "StartTime": "1400",
                "WeekText": "Every Week",
                "colorIndex": 0,
              },
              Object {
                "ClassNo": "ST5",
                "DayText": "Thursday",
                "EndTime": "1600",
                "LessonType": "Tutorial",
                "ModuleCode": "PC2132",
                "ModuleTitle": "",
                "StartTime": "1500",
                "WeekText": "Every Week",
                "colorIndex": 0,
              },
            ],
          ],
          "Wednesday": Array [
            Array [
              Object {
                "ClassNo": "ST1",
                "DayText": "Wednesday",
                "EndTime": "1100",
                "LessonType": "Tutorial",
                "ModuleCode": "PC4241",
                "ModuleTitle": "",
                "StartTime": "1000",
                "WeekText": "Every Week",
                "colorIndex": 1,
              },
              Object {
                "ClassNo": "ST2",
                "DayText": "Wednesday",
                "EndTime": "1400",
                "LessonType": "Tutorial",
                "ModuleCode": "PC2132",
                "ModuleTitle": "",
                "StartTime": "1300",
                "WeekText": "Every Week",
                "colorIndex": 0,
              },
              Object {
                "ClassNo": "ST3",
                "DayText": "Wednesday",
                "EndTime": "1700",
                "LessonType": "Tutorial",
                "ModuleCode": "PC4241",
                "ModuleTitle": "",
                "StartTime": "1600",
                "WeekText": "Every Week",
                "colorIndex": 1,
              },
            ],
          ],
        }
      }
      onModifyCell={[Function]}
      showTitle={false}
    />
  </div>
</li>
`;

exports[`VenueDetailRow it displays an anchor tag when not expanded 1`] = `
<li
  className="venueDetailRow"
>
  <a
    href="/venues/CQT%2FSR0622"
    onClick={[Function]}
  >
    <h4>
      CQT/SR0622
    </h4>
  </a>
</li>
`;
